Request for comments: rule colouring support

kmilner opened the pull request:
These commits introduce two changes: 

1. A way of manually specifying the colour a rule should be rendered in generated graph. This is done with an action `__color('RRGGBB')` or `__colour('RRGGBB')` included in the rule, where `RRGGBB` is a 24-bit colour in hexadecimal. (Currently, `#` is not allowed in string literals, or I would have had that in front too)

2. To support this, facts identifiers can now begin with `_`, and the facts in rule actions beginning with `__` (double underscore) are hidden when printing the rule and in the dot output.

Since identifiers beginning with `_` were previously unsupported, this should not introduce any incompatibilities. However, I would like feedback on the format or potential alternatives, as well as any other ideas for useful rendering annotations beyond colour.

(Separately, if some way of specifying hidden actions gets merged, we may want to switch the diff code to use the same thing rather than the current method of manually reconstructing diff annotations to compare against actions to hide them, as is done for example at Rule.hs#L885. I don't know the diff code very well though)

kmilner commented:
And of course, an example image!
![register_pk](https://cloud.githubusercontent.com/assets/8351604/24852654/3978036a-1dd0-11e7-8e56-059fc84eb067.png)

kmilner commented:
Here's an alternative where the font colour is picked according to a luminance formula that might well be made up. White font with antialiasing ends up being quite hard to read, but at least this would let people set the colour to black (though that looks bad for other reasons). Which is preferred?
![register_pk-3](https://cloud.githubusercontent.com/assets/8351604/24853508/3804382a-1dd3-11e7-810d-577838304ad3.png)

rkunnema commented:
Hi!

This is a useful feature for SAPIC to have, since we can mark protocol parties with different colours (protocol parties could be designated by let-expressions). At the first glance it would not collide with the action names SAPIC is using. So I am all for it.

Cheers, Robert

jdreier commented:
I like the idea of adding coulours and having a more transparent way of hiding facts, but allowing any identifier (not only facts) to start with _ will probably lead to some bugs, in particular in the equivalence part where we rely on underscores to generate unique names. I guess this could be avoided by distinguishing fact identifiers and other identifiers (variables, functions etc.).

kmilner commented:
Hi all,

For another feature I'm looking at adding (prioritisation of particular rule premises), I'd like to sort out a standardized way to annotate rules. I agree that having 'special' actions like this is probably not the best way, other than being easy to implement.

Would it be better to use a similar thing to the existing lemma annotations, i.e., 
`rule RuleName [colour=00FF00]: [ ] --[ ]-> []`? I don't think that would be too hard to implement, and then it would be consistent with lemmas.

kelnage commented:
In many ways I prefer the annotations (a standard) over the actions (convention) - but I guess this would likely prevent @rkunnema from colouring the rules based on the protocol parties?

jdreier commented:
Rule annotations seems to be a cleaner way of doing it, and I don't think this would be a problem for SAPIC. The colours have to be generated anyway when creating the rule, and the rest is pure syntax.

kmilner commented:
I have updated my branch to add in rule attribute parsing and printing, set up so that it will be easy to add additional rule annotations later on (and removed the previous `__action()` annotation method used above). 

I also added automatic font colour (white or black) based on the rule colour.

Let me know what you think.

cascremers commented:
@jdreier @rsasse how do you feel about the updated version? I'm okay with this and would be happy to sign off.

rsasse commented:
This makes sense to me. For a user that does not use such annotation nothing changes, and for those that do want to use it, they get additional visibility in their graphs. Sounds good!

jdreier commented:
I am fine with it, too. We should try the regression tests and then accept!

rsasse commented:
Indeed, regression is needed and case-studies/ake/dh/UM_three_pass_combined_fixed_analyzed.spthy probably will have to be updated as usual.

kmilner commented:
One thing, looking at this code again, I should improve the typing (`RuleColor` should really contain a colour type instead of a string, for example). I will do that now.

kmilner commented:
Alright, sorry for the delay. The colour is now stored as an RGB type, and the parser is now a bit more lenient about how you format your colour codes. It accepts 
`(color|colour)=('(#)?RRGGBB'|(#?)RRGGBB)`
that is, you can spell color however you like, single quote your colorcode if you want, and prefix it with a `#` if you want. It also checks that it gets correctly parsed by `hexToRGB`, otherwise fails with an error.

I have not yet run regression tests, I will get those going (assuming `make case-studies` and a diff is the procedure there).

One thing I noticed---and this also applies to lemma attributes---when the parser fails to grab an attribute it doesn't fail very gracefully: rather than trying to parse the rest of the list (or trashing it) and carrying on, it currently just fails out of loading the file entirely. But this is minor and a separate issue if anyone prefers it work differently.

rsasse commented:
For running the regression you want to do
rm -rf case-studies
make case-studies
diff -r case-studies case-studies-regression
and hopefully all the difference you see is in the runtime of the analyses. The one file I mention above manages to be reordered on any minor change (no idea how), so as long as the number of proof steps does not change, it is fine to commit the changed version.

Thanks!

kmilner commented:
@rsasse That would be an excellent thing to add to `CONTRIBUTING.md` also :)

rsasse commented:
Hm, yes and no. 

Yes, it would be nice if whoever submits a pull request did that regression check. No, because they could just claim they did it, and so whoever (Cas, Jannik, or myself) accepts the PR must run it themselves anyway. In this case, as Cas is the one who will accept, and he knows where your office is, this is not a problem. ;-) But it should be documented somewhere anyway, that's true. So, I just did it and committed directly.

kmilner commented:
Case studies diff attached. I'm not sure I believe your comment about runtime changes on the order of 1%, though I ran this on a pretty recent server.
[case-studies-diff](https://github.com/tamarin-prover/tamarin-prover/files/1159263/case-studies-diff.txt)

rsasse commented:
This looks good. Note that the 1% speed difference is with respect to a previous run on the same machine! Currently the "ground truth" happens to be my (old) laptop. ;)

To summarize, this diff is totally fine and expected.

cascremers commented:
The change is fine, but for the record, the case studies diff file linked above by @kmilner is not meaningful, since it compares two different versions running on two different machines. In other tests, in which we compared this change to the base version on the same machine, the difference is indeed negligible. This  is what we had expected, but please ignore the file linked above.

cascremers commented:
@kmilner I would be great if you could add a paragraph to the manual to document this feature.
